mtcpclient/groupofconnections.go

```diff
@@ -29,23 +48,89 @@ func (gc GroupOfConnections) String() string {
 		nTotal, nDialing, nEstablished, nClosed, nError, nNotInitiated)
 }
 
-func (gc GroupOfConnections) isIn(status tcpclient.ConnectionStatus) bool {
-	for _, item := range gc {
+func (gc GroupOfConnections) containsAConnectionWithStatus(status tcpclient.ConnectionStatus) bool {
+	for _, item := range gc.connections {
 		if item.GetConnectionStatus() == status {
 			return true
 		}
 	}
 	return false
 }
 
+// PendingConnections retuns True if at least one connection is being processed
 func (gc GroupOfConnections) PendingConnections() bool {
-	return gc.isIn(tcpclient.ConnectionNotInitiated) || gc.isIn(tcpclient.ConnectionDialing)
+	for _, connection := range gc.connections {
+		if connection.PendingToProcess() {
+			return true
+		}
+	}
+	return false
 }
 
+// AtLeastOneConnectionInError returns True is at least one connection establishment failed
 func (gc GroupOfConnections) AtLeastOneConnectionInError() bool {
-	return gc.isIn(tcpclient.ConnectionError)
+	for _, connection := range gc.connections {
+		if connection.WithError() {
+			return true
+		}
+	}
+	return false
+}
+
+func (gc GroupOfConnections) atLeastOneConnectionOK() bool {
+	for _, connection := range gc.connections {
+		if connection.WentOk() {
+			return true
+		}
+	}
+	return false
+}
+
+func (gc GroupOfConnections) pingStyleReport(typeOfReport string) (output string) {
+	var connectionsFiltered GroupOfConnections
+	switch typeOfReport {
+	case "successful":
+		connectionsFiltered = gc.getConnectionsThatWentWell()
+		output += "Response time stats for " + strconv.Itoa(len(connectionsFiltered.connections)) +
+			" successful connections min/avg/max/dev = " + printStats(connectionsFiltered.calculateMetricsReport())
+	case "errored":
+		connectionsFiltered = gc.getConnectionsThatWentBad()
+		output += "Time to error stats for " + strconv.Itoa(len(connectionsFiltered.connections)) +
+			" failed connections min/avg/max/dev = " + printStats(connectionsFiltered.calculateMetricsReport())
+	}
+	return output
+}
+
+func (gc GroupOfConnections) getConnectionsThatWentWell() (connectionsThatWentWell GroupOfConnections) {
```

Comment:
too much duplication between this and the next function. Maybe the check condition can be a function-argument and you only have something like "getConnectionsThatWent(functionArg())"

Reply:
I solved it another way, but we could find a better approach (using your approach I found some blocker)